### PR TITLE
fix crash on semaphoreCI caused by unaligned bitops in TypeFunction

### PR DESCRIPTION
**Pull request by ghost**
It is caused by https://github.com/dlang/dmd/pull/11461. Would be generated bitop that caused corruption when the flag container is not aligned by 4 bytes.

### CODE REVIEW

**Comment by NilsLankila** [comment]
It looks like the problem is the bitops on non aligned address, i.e the TypeFunction.flags are located at `0x<somevalue>E`flags.

But why only on CircleCI ?

**Comment by UplinkCoder** [comment]
Make funcflags an int

**Comment by NilsLankila** [comment]
yeah but instance size is exactly 128... so it waste possibly the refact. (I dont remember what was the old size).

**Comment by NilsLankila** [comment]
DMD generates bitop instructions that works on 32 bit or 64 bit registers, so what happens is that the two bytes following the instance are sometimes corrupted... but again, why only on SemaphoreCI ?

**Comment by MoonlightSentinel** [comment]
Fine with merging this but it seems more like a workaround because
- it didn't happen for your initial PR (suggesting there was another trigger)
- it crashes in `isnothrow(bool)`, the unaligned read and bitops inside of `isnothrow()` pass

**Comment by NilsLankila** [comment]
It didn't happen because SemaphoreCI was not used anymore. It's been fixed only a few days after.

**Comment by NilsLankila** [comment]
But otherwise I agree that this is a workaround. Can be the store operation of isnothrow(bool), can be a backend bug, and the question of  why only on semaphoreci remains.

**Comment by kinke** [comment]
Seems like a backend bug, although the base type of `FunctionFlag` is a `int` and should probably be a ushort instead. Btw, why is the field a ushort and not simply th enum directly?

**Comment by NilsLankila** [comment]
The enum is is an `int` because of deprecation on `~`. The field is an ushort and there's a static test that verify that it stay small enough.

**Comment by kinke** [comment]
Ah, I see. I think the manual ushort 'optimization' is overkill to potentially save a few bytes - the other enums used as TypeFunction fields (LINK, TRUST, PURE) are all ints and stored as ints, although they only have very few members. - I think these different alignments - 2 for the ushort field, 4 for the enum value - is causing this.

File: src/dmd/astbase.d [hunk 1]
```diff
@@ -3986,8 +3986,8 @@ struct ASTBase
         TRUST trust;                // level of trust
         Expressions* fargs;         // function arguments
         PURE purity = PURE.impure;
-        byte inuse;
         ushort funcFlags;
+        byte inuse;
```
Comment (NilsLankila) [inline src/dmd/astbase.d:3990]: a better solution is to put this in the bitset

**Comment by MoonlightSentinel** [comment]
> It didn't happen because SemaphoreCI was not used anymore. It's been fixed only a few days after.

Right but it was green for many other PR's after your PR was merged 21 days ago.
Manually bisecting found that https://github.com/dlang/dmd/pull/11593#issuecomment-676339595 started after df742683dbe26c20c08d0cc1817be83bc7b9e7f9, allthough i hardly doubt that this commit is the root cause.


> [...} and the question of why only on semaphoreci remains.

SemaphoreCI is the only CI doing self-hosted tests IIRC.

File: src/dmd/astbase.d (hunk 1 above)
Comment (NilsLankila) [reply src/dmd/astbase.d:3990]: unfortunately this is one of the rare `inuse` variable that can have 2 has value.

**Comment by NilsLankila** [comment]
> Right but it was green for many other PR's after your PR was merged 21 days ago.

Another explanation is that the parent commit of the PRs coming after mine was older than the commit it  contained, so for a few weeks everything looked okay, until new PR come that was, this time, up to date with ~master.

**Review by MoonlightSentinel — APPROVED** [review]
This change is trivial and unblocks the CI. Can easily be reverted once we figure out the root cause

**Comment by MoonlightSentinel** [comment]
> Another explanation is that the parent commit of the PRs coming after mine was older than the commit it contained, so for a few weeks everything looked okay, until new PRs come that was, this time, up to date with ~master.

Just checked, SemaphoreCI doesn't merge back to master (as done e.g. on the auto-tester). That might be an explaination why this slipped through.

**Comment by UplinkCoder** [comment]
The cause for this is almost surely a codegen bug in dmd.
I assume if it were compiled with ldc this problem would go away.

**Comment by NilsLankila** [comment]
>  this problem would go away.

That was exactly the case on semaphore. same test with 4 different compilers. only DMD x86_64 crashed.

**Comment by ibuclaw** [comment]
Is there a bugzilla report for the crash? FYI @WalterBright.

**Pull request merged**